File: Binary_Trees/105-construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.py
    # Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# Definition for a binary tree node.
class Solution:
    def buildHelper(self, preorder, ps, pl, inorder, is_, il, index):
        # Base case: if the start index is greater than the end index, return None
        if ps > pl or is_ > il:
            return None
        
        # Root node is the first element in the preorder traversal
        root_val = preorder[ps]
        root = TreeNode(root_val)
        
        # Find the index of the root value in the inorder traversal using the index map
        inorder_index = index[root_val]
        
        # Calculate the number of elements in the left subtree
        left_subtree_size = inorder_index - is_

        # Recursively build the left subtree
        root.left = self.buildHelper(preorder, ps + 1, ps + left_subtree_size, inorder, is_, inorder_index - 1, index)
        
        # Recursively build the right subtree
        root.right = self.buildHelper(preorder, ps + left_subtree_size + 1, pl, inorder, inorder_index + 1, il, index)
        
        return root

    def buildTree(self, preorder: list[int], inorder: list[int]) -> TreeNode:
        # Create a map to store the index of each value in the inorder traversal
        index = {val: idx for idx, val in enumerate(inorder)}
        
        # Call the helper function to build the tree
        return self.buildHelper(preorder, 0, len(preorder) - 1, inorder, 0, len(inorder) - 1, index)

File: Binary_Search_Trees/105-construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildHelper(vector<int>& preorder, int ps, int pl, vector<int>& inorder, int is_, int il, unordered_map<int, int>& indexMap) {
        // Base case: if the start index is greater than the end index, return NULL
        if (ps > pl ) {
            return nullptr;
        }
        
        // Root node is the first element in the current range of the preorder traversal
        int rootVal = preorder[ps];
        TreeNode* root = new TreeNode(rootVal);
        
        // Find the index of the root value in the inorder traversal using the index map
        int inorderIndex = indexMap[rootVal];
        
        // Calculate the size of the left subtree
        int leftSubtreeSize = inorderIndex - is_;
        
        // Recursively build the left subtree
        root->left = buildHelper(preorder, ps + 1, ps + leftSubtreeSize, inorder, is_, inorderIndex - 1, indexMap);
        
        // Recursively build the right subtree
        root->right = buildHelper(preorder, ps + leftSubtreeSize + 1, pl, inorder, inorderIndex + 1, il, indexMap);
        
        return root;
    }

    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        // Create a map to store the index of each value in the inorder traversal
        unordered_map<int, int> indexMap;
        for (int i = 0; i < inorder.size(); i++) {
            indexMap[inorder[i]] = i;
        }

        // Call the helper function to build the tree
        return buildHelper(preorder, 0, preorder.size() - 1, inorder, 0, inorder.size() - 1, indexMap);
    }
};